service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureNetworkInterfaceStep.java

```diff
@@ -108,6 +116,46 @@ public StepResult doStep(FlightContext context) throws InterruptedException, Ret
     return StepResult.getStepResultSuccess();
   }
 
+  NetworkSubnetPair getExistingNetworkResources(
+      AzureCloudContext azureCloudContext, NetworkManager networkManager) {
+
+    // we are considering the network id to be optional
+    if (resource.getNetworkId() == null) {
+      return getNetworkResourcesFromLandingZone(azureCloudContext, networkManager);
+    }
+
+    final ControlledAzureNetworkResource networkResource =
+        resourceDao
+            .getResource(resource.getWorkspaceId(), resource.getNetworkId())
+            .castByEnum(WsmResourceType.CONTROLLED_AZURE_NETWORK);
+
+    return NetworkSubnetPair.createNetworkSubnetPair(
+        networkManager,
+        azureCloudContext.getAzureResourceGroupId(),
+        networkResource.getNetworkName(),
+        networkResource.getSubnetName());
+  }
+
+  private NetworkSubnetPair getNetworkResourcesFromLandingZone(
+      AzureCloudContext azureCloudContext, NetworkManager networkManager) {
+
+    final String lzId = landingZoneApiDispatch.getLandingZoneId(azureCloudContext);
+
+    ApiAzureLandingZoneDeployedResource lzResource =
+        landingZoneApiDispatch
+            .listSubnetsWithParentVNetByPurpose(
```

Comment:
Just a thought here: we might want to add REST API for the landing zone subnets ../{lz_id}/networkSubnets/{SubnetResourcePurpose}, or for getting resources by purpose ../{lz_id}/resources/{LandingZonePurpose} for better performance.
List all resources is an expensive API because it makes calls resource manager to retrieve by tag (N+1) times, where N is a number of purposes, and M times to get resource by Id, where M is a number of VNets in landing zone => N+M+1.
The scenario, however, is to get subnets with exact purpose that is 1 call to get VNets by tag and max(M) times (if each VNet has the subnet for a purpose specified) to get by Id => 1+max(M).

Reply:
I agree. We can optimize this as a follow up. The LZ manager already has a method for querying VNets/Subnets by purpose. We just need to expose it in the service etc..